weaver/wps_restapi/jobs/utils.py

```diff
@@ -496,69 +631,445 @@ def get_job_results_response(job, container, headers=None):
+    :return:
+    """
+    # FIXME: implement (https://github.com/crim-ca/weaver/pull/548)
+    #   for .../results/{id} transform might need to force 'Prefer' over job preference
+    #   (explicitly request value/link contrary to resolved results/mode from the job)
+
+    is_ref = bool(get_any_value(result, key=True, file=True, data=False))
+    out_data = get_any_value(result, file=is_ref, data=not is_ref)
+    out_mode, out_fmt = get_job_output_transmission(job, output_id, is_ref)
+    output_format = output_format or out_fmt
+    if out_mode == ExecuteTransmissionMode.REFERENCE:
+        link = make_result_link(job, result, output_id, out_mode, output_format, settings=settings)
+        headers.extend([("Link", link[0])])
+        return HTTPNoContent(headers=headers)
+
+    # convert value as needed since reference transmission was not requested/resolved
+    out_headers = {}
+    if is_ref:
+        output_mode = ExecuteTransmissionMode.VALUE
+        out_headers, out_data = generate_or_resolve_result(
+            job,
+            result,
+            output_id,
+            output_id,
+            output_mode,
+            output_format,
+            settings=settings,
+        )
+        headers.update(out_headers)
+
+    ctype = out_headers.get("Content-Type")
+    if not ctype:
+        ctype = get_field(result, "mediaType", search_variations=True, default=ContentType.TEXT_PLAIN)
+    c_enc = cast("AnyContentEncoding", headers.get("Content-Encoding") or "UTF-8")  # type: AnyContentEncoding
+    out_data = data2str(out_data)
+    out_data = ContentEncoding.encode(out_data, c_enc)
+    return HTTPOk(body=out_data, content_type=ctype, charset=c_enc, headers=headers)
+
+
+def get_job_results_document(job, results, *, settings):
+    # type: (Job, ExecutionResults, Any, SettingsType) -> ExecutionResults
+    """
+    Generates the :term:`Job` results document response from available or requested outputs with necessary conversions.
+
+    Removes nested literal value definitions if qualified value representation is not needed.
+    Qualified value representation is not needed if no other field than ``value`` is provided with the literal data,
+    or when the specified :term:`Media-Type` is simply the plain text default for data literals.
+    The simplification is applied for both literals on their own and nested array of literals.
+    However, when processing an array, the qualified value representation is preserved if any of the items requires
+    the explicit mention of another :term:`Media-Type` than plain text, to return a consistent structure.
+
+    Uses the :paramref:`job` definition and submitted ``headers``
+
+    .. warning::
+        This function assumes that schema deserialization was applied beforehand.
+        Therefore, it will not attempt matching every possible combination of the results representation.
+    """
+
+    def make_result(result, result_id, output_id):
```

Comment:
Why have it inside `get_job_results_document` and not at the class level like `make_result_link`

Reply:
It is a nested function. It is relevant only within the context of the document response. This way, I avoid misusing this function in other contexts and potentially introducing side effects that only `get_job_results_document` can assume.